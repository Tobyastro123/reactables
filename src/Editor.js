import React from 'react'

var EditorConstants = require('./EditorConstants');

class Editor extends React.Component {
  
}

export default Editor
