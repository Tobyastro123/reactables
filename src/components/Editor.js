import React from 'react';
import {EditorException} from './Exceptions';
import editorStyles from '../css/editor.css';
import classNames from 'classnames/bind';
import superagent from 'superagent';

var CommonConstants = require('./CommonConstants');
var EditorConstants = require('./EditorConstants');
var Lang = require('./Lang');

class Editor extends React.Component {
  constructor(props)
  {
    super(props);
    if (typeof props.editor.fields === EditorConstants.UNDEFINED) {
      throw new EditorException('You should define "fields" option.');
    }
    if (props.editor.fields.length === 0) {
      throw new EditorException('You should define at least one field in "fields" option.');
    }
    this.lang = Lang[props.lang];
    this.state = {
      dataIndices: {},
      popup_title: this.lang.gte_editor_popupheader_create,
      popup_button: this.lang.gte_editor_sendbtn_create
    }
    this.setFields(props);
    this.setDataIndices(props);
  }

  setDataIndices(props)
  {
    let cols = props.columns;
    this.dataIndices = [];
    cols.map((column, index) => {
      this.dataIndices[column.name] = '';
    });
  }

  setFields(props)
  {
    let fields = [];
    if (props.action === EditorConstants.ACTION_CREATE) {
      fields = this.setCreateFields(props.editor.fields);
    } else if (props.action === EditorConstants.ACTION_EDIT) {
      fields = this.setEditFields(props.editor.fields);
    } else if (props.action === EditorConstants.ACTION_DELETE) {
      fields = this.setDeleteFields(props.selectedRows);
    }
    this.fields = fields;
  }

  setCreateFields(editorFields)
  {
    let fields = [];
    editorFields.map((object, index) => {
      fields[index] = this.getFieldByType(index, object);
    });
    return fields;
  }

  setEditFields(editorFields)
  {
    let fields = [];
    editorFields.map((object, index) => {
      fields[index] = this.getFieldByType(index, object);
    });
    return fields;
  }

  setDeleteFields(items)
  {
    let fields = [], lastId = 0;
    this.state.dataIndices = this.props.selectedIds;
    this.props.selectedIds.map((object, index) => {
      fields[index] = <input key={index} type="hidden" data-value={object} name="ids[]" value={object} />;
      lastId = index;
    });
    fields.push(<div key={++lastId} className="gte_msg">Are You sure You wish to delete {items.length} row(s)?</div>);
    return fields;
  }

  onChange(e)
  {
    this.setState({
      dataIndices: Object.assign({}, this.state.dataIndices, {
        [e.target.name]: e.target.value
      })
    })
  }

  fileUpload()
  {
    if (this.filesInput.files !== CommonConstants.UNDEFINED) {
      const { ajaxFiles } = this.props.editor;
      let formData = new FormData();
      const files = this.filesInput.files;
      for (var key in files) {
        // check if this is a file:
        if (files.hasOwnProperty(key) && files[key] instanceof File) {
            formData.append(key, files[key]);
        }
      }

      superagent.post(ajaxFiles)
      .send(formData)
      .end((err, response) => {
        if(err) {
            console.log('Error has accured while uploading files: ');
            console.log(err);
        } else if(response.ok) {
            //this was successful, handle it here
        }
      });
    }
  }

  getFieldByType(index, object)
  {
    const { dataIndices } = this.state;
    var fieldType = object.type,
            fieldName = object.name,
            fieldLabel = object.label,
            action = this.props.action,
            fieldValue = '';
    // settting attrs
    var attributes = [];
    if (typeof object.attrs !== CommonConstants.UNDEFINED) {
      var fieldOpts = object.attrs;
      for (var opt in fieldOpts) {
        for (var attr in fieldOpts[opt]) {
          attributes[attr] = fieldOpts[opt][attr];
        }
      }
    }

    if (action === EditorConstants.ACTION_EDIT) {
      if (typeof dataIndices[fieldName] !== CommonConstants.UNDEFINED) {
        fieldValue = dataIndices[fieldName];
      } else {
        fieldValue = this.props.fieldsEdit[fieldName];
      }
    } else if (action === EditorConstants.ACTION_CREATE && typeof dataIndices[fieldName] !== CommonConstants.UNDEFINED) {
      fieldValue = dataIndices[fieldName];
    }

    let i = 0,
    htmlFields = [];
    switch (fieldType) {
      // uncontrolled input element, so we can put value here
      case EditorConstants.TYPE_TEXT:
      case EditorConstants.TYPE_HIDDEN:
      case EditorConstants.TYPE_EMAIL:
      case EditorConstants.TYPE_PASSWORD:
        htmlFields[i] = <div className="gte_editor_fields">
          <label className="gte_label" htmlFor={fieldName}>{(fieldType !== EditorConstants.TYPE_HIDDEN) ? fieldLabel : null}</label>
          <div className={editorStyles.gte_field}>
            <input onChange={this.onChange.bind(this)} {...attributes} id={fieldName} type={fieldType} name={fieldName}
              value={fieldValue} />
          </div>
          <div className="clear"></div></div>;
        break;
      case EditorConstants.TYPE_COLOR:
      case EditorConstants.TYPE_DATE:
      case EditorConstants.TYPE_DATETIME:
      case EditorConstants.TYPE_NUMBER:
      case EditorConstants.TYPE_RANGE:
      case EditorConstants.TYPE_SEARCH:
      case EditorConstants.TYPE_TIME:
      case EditorConstants.TYPE_TEL:
      case EditorConstants.TYPE_URL:
      case EditorConstants.TYPE_MONTH:
      case EditorConstants.TYPE_WEEK:
        htmlFields[i] = <div className="gte_editor_fields"><label className="gte_label" htmlFor={fieldName}>{fieldLabel}</label><div className={editorStyles.gte_field}><input onChange={this.onChange.bind(this)} {...attributes} id={fieldName} type={fieldType} name={fieldName} /></div><div className="clear"></div></div>;
        break;
      case EditorConstants.TYPE_FILE:
        htmlFields[i] = <div className="gte_editor_fields">
          <label className="gte_label" htmlFor={fieldName}>{fieldLabel}</label>
            <div className={editorStyles.gte_field}>
              <input ref={(input) => { this.filesInput = input; }} {...attributes} id={fieldName} type={fieldType} name={fieldName} />
            </div>
            <div className="clear"></div>
          </div>;
        break;
      case EditorConstants.TYPE_TEXTAREA:
        htmlFields[i] = <div className="gte_editor_fields">
          <label className="gte_label" htmlFor={fieldName}>{fieldLabel}</label>
          <div className={editorStyles.gte_field}>
            <textarea onChange={this.onChange.bind(this)} {...attributes} id={fieldName} name={fieldName} value={fieldValue}></textarea>
          </div>
          <div className="clear"></div></div>;
        break;
      case EditorConstants.TYPE_SELECT:
        var values = object.values;
        var options = [], val = '';
        for (var k in values) {
          for (var key in values[k]) {
            val = values[k][key].trim();
            options[k] = <option key={key} value={key} data-value={val.toLowerCase()}>{val}</option>;
          }
        }
        htmlFields[i] = <div className="gte_editor_fields">
          <label className="gte_label" htmlFor={fieldName}>{fieldLabel}</label>
          <div className={editorStyles.gte_field}>
            <select value={fieldValue} onChange={this.onChange.bind(this)} {...attributes} id={fieldName} name={fieldName}>{options}</select>
          </div>
          <div className="clear"></div></div>;
        break;
      case EditorConstants.TYPE_CHECKBOX:
      case EditorConstants.TYPE_RADIO:
        var values = object.values;
        var options = [], val = '',
                //@fixme regexp to remove ex: [3] etc
                id = fieldName.replace('[]', '');
        for (var k in values) {
          for (var key in values[k]) {
            val = values[k][key].trim();
            options[k] = <label key={key} className="gte_label_text">
              <input defaultChecked={(val === fieldValue) ? 1 : 0 } onClick={this.onChange.bind(this)} {...attributes} id={id} type={fieldType} name={fieldName} data-value={val.toLowerCase()} value={key}/>{val}</label>;
          }
        }
        htmlFields[i] = <div className="gte_editor_fields"><label className="gte_label">{fieldLabel}</label><div className={editorStyles.gte_field}>{options}</div><div className="clear"></div></div>;
        break;
    }
    return htmlFields;
  }

  triggerBefore(type)
  {
    const { tableOpts } = this.props;
    // call triggerBefore if it has been set
    tableOpts.buttons.map((obj, idx) => {
      if (obj.extended === type && typeof obj.triggerBefore !== CommonConstants.UNDEFINED) {
          obj.triggerBefore();
      }
    });
  }

  triggerAfter(type)
  {
    const { tableOpts } = this.props;
    // call triggerAfter if it has been set
    tableOpts.buttons.map((obj, idx) => {
      if (obj.extended === type && typeof obj.triggerAfter !== CommonConstants.UNDEFINED) {
          obj.triggerAfter();
      }
    });
  }

  btnClicked(e)
  {
    e.persist(); // this is to avoid null values in this.props.editorUpdate(e, dataResp) call
    const { action, editorUpdate } = this.props;
    let ajaxUrl = this.props.editor.ajax, that = this;
    var dataResp = that.state.dataIndices;
    if (action === EditorConstants.ACTION_CREATE) {
      this.triggerBefore(EditorConstants.EDITOR_CREATE);
      this.fileUpload();
      fetch(ajaxUrl, {
      method: EditorConstants.HTTP_METHOD_POST,
      body: JSON.stringify(this.state.dataIndices)
      }).then(response => response.json()).then((data) => {
        dataResp['id'] = data['row']['id'];
        dataResp[CommonConstants.GT_ROW_ID] = data['row']['id'];
        editorUpdate(e, dataResp);
        this.triggerAfter(EditorConstants.EDITOR_CREATE);
      });
    } else if (action === EditorConstants.ACTION_EDIT) {
      this.triggerBefore(EditorConstants.EDITOR_EDIT);
      this.fileUpload();
      fetch(ajaxUrl, {
      method: EditorConstants.HTTP_METHOD_PUT,
      body: JSON.stringify(this.state.dataIndices)
      }).then(response => response.json()).then((data) => {
        editorUpdate(e, dataResp);
        this.triggerAfter(EditorConstants.EDITOR_EDIT);
      });
    } else if (action === EditorConstants.ACTION_DELETE) {
      this.triggerBefore(EditorConstants.EDITOR_REMOVE);
      fetch(ajaxUrl, {
        method: EditorConstants.HTTP_METHOD_DELETE,
        body: JSON.stringify(this.props.dataIndices)
      }).then(response => response.json()).then((data) => {
        // call editorUpdate method with passing all user-input values
        editorUpdate(e, dataResp);
        this.triggerAfter(EditorConstants.EDITOR_REMOVE);
      });
    }
  }

  stopPropagation(e)
  {
    e.stopPropagation();
  }

  render()
  {
    const {
      hidePopup,
      opacity,
      popupTitle,
      action,
      popupButton,
      active
    } = this.props;
    this.setFields(this.props);
    let editorClasses = classNames({
      gte_editor_popup: true,
      fade_in: active,
      fade_out: !active
    }),
    backgroundClasses = classNames({
      gte_popup_background:true,
      fade_in: active,
      fade_out: !active
    });
    return (
      <div>
      <div onClick={hidePopup} className={editorClasses}>
        <div className="gte_popup_container">
          <div className="gte_popup_container_wrapper">
            <div onClick={this.stopPropagation.bind(this)} className="gte_form_border_box">
              <div className="gte_form_fields">
                <div className="gte_header">
                  <div className="gte_editor_title">{popupTitle}</div>
                </div>
                <div className="gte_form_body">
                  <div className="gte_form_body_content">
                    <form id="gte_form" action="" method="post">
                      <div className="gte_form_content">
                        <div>
                          <input type="hidden" name="action" value={action}/>
                        </div>
                        <div>
                          {this.fields}
                        </div>
                      </div>
                    </form>
                  </div>
                </div>
                <div className="gte_footer">
                  <div className="gte_form_err"></div>
                  <div className="gte_form_buttons">
                    <button
                      id="gte_sent_btn"
                      className="btn"
                      data-action={action}
                      onClick={this.btnClicked.bind(this)}>{popupButton}</button>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div onClick={hidePopup} className={backgroundClasses}></div>
      </div>
    )
  }
}

export default Editor
