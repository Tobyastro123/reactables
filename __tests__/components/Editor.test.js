import React from 'react';
import renderer from 'react-test-renderer';
import Editor from '../../src/components/Editor.js'
import { shallow } from 'enzyme';

var editor = {
  ajax: 'http://gigatables.loc/editor.php',
  ajaxFiles: 'uploadFiles.php',
  struct: {
    buttons: ['top', 'bottom'] // buttons
  },
  fields: [
    {
      label: "ID",
      name: "id",
      type: 'hidden'
    },
    {// an example of using select - automatically selected if matches with data in table column
      label: "Types:",
      name: "types[]",
      values: [// if select,checkbox,radio etc types - need this pre-set structure of values
        {'key1': 'val1'},
        {'key2': 'val2'}
      ],
      type: 'checkbox', // select,checkbox,radio
//              attrs: [
//                {'multiple':true}
//              ]
    },
    {
      label: "Article title:",
      name: "title",
      type: 'text', // default, other: password, file, select, multiselect etc
      attrs: [
        {'pattern': '^[A-Za-z0-9_]+$'}
      ]
    },
    {
      label: "Description:",
      name: "desc",
      type: 'textarea'
    },
    {
      label: "Date Time:",
      name: "date",
      type: 'date'
    },
    {
      label: "Image:",
      name: "image",
      type: 'file'
    },
  ]
};

var settings = {
 struct: {
   search: ['top', 'bottom'],
   rowsSelector: ['desc', 'top', 'bottom'],
   pagination: ['bottom']
 },
 lang: 'en', // english default
 perPageRows: [25, 50, 100, 200],
 defaultPerPage: 100,
 ajax: 'http://gigatables.loc/gigatables.php',
 // ajaxAutoloadData: true, // default false
 // ajaxAutoloadPeriod: 8, // sec
 requestType: 'GET',
 columns: [
   {// include all defaults
     data: "id",
     sortable: true, // true by defualt
     visible: true, // true by defualt
     searchable: true, // true by defualt
     discreteSearch: true, // false by default
     discreteSearchValue: function (title) {
       return 'Search by field - ' + title;
     }
   },
   {
     data: "title",
     cISearch: true // default false
   },
   {
     data: "desc",
     sortable: false,
     discreteSearch: true,
     discreteCISearch: true // default false
   },
   {
     data: "date",
     searchable: false
   },
   {
     data: "info"
   },
   {data:"field1"},
   {data:"field2"},
   {data:"field3", visible: false}
 ],
 columnOpts: [
   {
     render: (data, row, type) => (
       <div>
         <form method="post" action="">
          <input type="hidden" name="action" value="forms" />
          <input type="hidden" name="id" value={row.id} />
          <div>{data}</div>
        </form>
      </div>),
     target: 'title' // provide data column index to match opts
   },
   {
     render: (data, row, type) => (
        <div>
          <div>{data}</div>
          <div>{(row.id % 2 === 0) ? '**' : '*'}</div>
        </div>
       ),
     target: 'date'
   }
 ],
 tableOpts: {
   buttons: [
     {extended: "editor_create", editor: editor, triggerAfter: (function () {
         console.log('after create');
       }), triggerBefore: (function () {
         console.log('before create');
       })},
     {extended: "editor_edit", editor: editor, triggerBefore: (function () {
       console.log('before edit');
     })},
     {extended: "editor_remove", editor: editor, triggerAfter: (function () {
         console.log('after del');
       })}
   ],
   buttonsPosition: ['top', 'bottom'],
   theme: 'std'
 }
};

describe("Editor", () => {

beforeEach(() => {
  global.editorUpdate = jest.fn().mockImplementation((e, dataResp) => {

  });
  global.fetch = jest.fn().mockImplementation(() => {
      var p = new Promise((resolve, reject) => {
        resolve({
          ok: true,
          Id: '123',
          json: function() {
            return {"rows":[
              {
                "GT_RowId":123,
                "title":"Test Bar 123st row",
                "id":123,
                "desc":"Lorem Ipsum is simply dummy Bar 7196 text of the printing and typesetting",
                "info":"some info some info some info some info",
                "date":"13:23:43 02:04:2017",
                "field1":123,
                "field2":1357,
                "field3":12468
              }
            ]};
          }
        });
      });
      return p;
  });
});

it('renders Editor correctly', () => {
  const tree = renderer.create(
    <Editor
    lang="en"
    action="create"
    countRows="229"
    fromRow="0"
    page="1"
    perPage="50"
    active={true}
    editor={editor}
    columns={editor.fields}
    struct={settings.struct}
    tableOpts={settings.tableOpts}
    selectedRows={[1]}
    selectedIds={[187]}
    fieldsEdit={{}} />
  ).toJSON();
  expect(tree).toMatchSnapshot();

  const obj = shallow(
    <Editor
    lang="en"
    countRows="229"
    fromRow="0"
    page="1"
    perPage="50"
    action="delete"
    active={true}
    editor={editor}
    columns={editor.fields}
    struct={settings.struct}
    tableOpts={settings.tableOpts}
    selectedRows={[1]}
    selectedIds={[187]}
    fieldsEdit={{}} />
  );

  obj.instance().setDataIndices({
    columns: settings.columns
  });

  obj.instance().setFields({
    action: 'create',
    editor: editor
  });

  obj.instance().setFields({
    action: 'edit',
    editor: editor
  });

  obj.instance().setFields({
    action: 'delete',
    editor: editor,
    selectedRows: [1, 2, 3]
  });

  obj.instance().triggerBefore({
    type: 'editor_create'
  });

  obj.instance().triggerAfter({
    type: 'editor_remove'
  });

  obj.instance().onChange({
    target: {
      name: 'foo',
      value: 'bar'
    }
  });

  obj.instance().stopPropagation({
    stopPropagation: () => {}
  });

  obj.instance().btnClicked({
    persist: () => {}
  });

  const objEdit = shallow(
    <Editor
    lang="en"
    countRows="229"
    fromRow="0"
    page="1"
    perPage="50"
    action="edit"
    active={true}
    editor={editor}
    columns={editor.fields}
    struct={settings.struct}
    tableOpts={settings.tableOpts}
    selectedRows={[1]}
    selectedIds={[187]}
    fieldsEdit={{}} />
  );

  objEdit.instance().btnClicked({
    persist: () => {}
  });

});

});
