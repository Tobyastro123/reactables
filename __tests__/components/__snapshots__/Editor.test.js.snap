exports[`test renders Editor correctly 1`] = `
<div>
  <div
    className="gte_editor_popup fade_in"
    onClick={undefined}>
    <div
      className="gte_popup_container">
      <div
        className="gte_popup_container_wrapper">
        <div
          className="gte_form_border_box"
          onClick={[Function]}>
          <div
            className="gte_form_fields">
            <div
              className="gte_header">
              <div
                className="gte_editor_title" />
            </div>
            <div
              className="gte_form_body">
              <div
                className="gte_form_body_content">
                <form
                  action=""
                  id="gte_form"
                  method="post">
                  <div
                    className="gte_form_content">
                    <div>
                      <input
                        name="action"
                        type="hidden"
                        value={undefined} />
                    </div>
                    <div />
                  </div>
                </form>
              </div>
            </div>
            <div
              className="gte_footer">
              <div
                className="gte_form_err" />
              <div
                className="gte_form_buttons">
                <button
                  className="btn"
                  data-action={undefined}
                  id="gte_sent_btn"
                  onClick={[Function]} />
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className="gte_popup_background fade_in"
    onClick={undefined} />
</div>
`;
